teps/0044-decouple-task-composition-from-scheduling.md

```diff
@@ -0,0 +1,954 @@
+* Requires creating new Tasks to leverage the reuse (maybe embedded specs negate this?)
+* Doesn't help with parallel use cases
+
+### Remove distinction between Tasks and Pipelines
+
+In this version, we try to combine Tasks and Pipelines into one thing; e.g. by getting rid of Pipelines and adding all
+the features they have to Tasks, and by giving Tasks the features that Pipelines have which they do not have.
+
+Things Tasks can do that Pipelines can't:
+* Sidecars
+* Refer to images (including args to images like script, command, args, env....)
+
+Things Pipelines can do that Tasks can't:
+* Create DAGs, including running in parallel
+* Finally
+* When expressions
+
+For example, say our new thing is called a Foobar:
+
+```yaml
+kind: Foobar
+metadata:
+  name: git-clone
+spec:
+ workspaces:
+  - name: source-code
+ foobars:
+ - name: get-source
+   steps: # or maybe each FooBar can only have 1 step and we need to use runAfter / dependencies to indicate ordering?
+    - name: clone
+      image: gcr.io/tekton-releases/github.com/tektoncd/pipeline/cmd/git-init:v0.21.0
+      script: <script here>
+   workspaces:
+   - name: source-code
+     workspace: source-code
+ finally:
+ # since we merged these concepts, any Foobar can have a finally
+```
+
+```yaml
+kind: Foobar
+metadata:
+  name: build-test-deploy
+spec:
+ workspaces:
+  - name: source-code
+  - name: test-results
+ foobars:
+ - name: get-source
+   workspaces:
+   - name: source-code
+     workspace: source-code
+   foobarRef: # foobars could have steps or foobarRefs maybe?
+     name: git-clone # uses our foobar above
+ - name: run-unit-tests
+   runAfter: get-source
+   steps:
+  - name: unit-test
+    image: docker.io/library/golang:$(params.version)
+    script: <script here>
+   workspaces:
+   - name: source-code
+     workspcae: source-code
+   - name: test-results
+     workspace: test-results
+ - name: upload-results
+   runAfter: run-unit-tests
+   foobarRef:
+     name: gcs-upload
+   params:
+   - name: location
+     value: gs://my-test-results-bucket/testrun-$(taskRun.name)
+   workspaces:
+   - name: data
+     workspace: test-results
+finally:
+- name: update-slack
+  params:
+  - name: message
+    value: "Tests completed with $(tasks.run-unit-tests.status) status"
+```
+
+We will need to add something to indicate how to schedule Foobars now that we won't have the convience of drawing the
+line around Tasks; we could combine this idea with one of the others in this proposal.
+
+(@ImJasonH has some ideas here that are much more aligned with generics in other langs, he might want to update the
+example here.)
+
+Pros:
+* Maybe the distinction between Tasks and Pipelines has just been making things harder for us
+* Maybe this is a natural progression of the "embedding" we already allow in pipelines?
+
+Cons:
+* Pretty dramatic API change
+* Foobars can contain Foobars can contain Foobars can contain Foobars
```

Comment:
Indeed. On the other hand, we could experiment with it almost *completely* independently of the current CRDs (`Pipeline`, `Task`) 👼🏼. Different api name, group and version, maybe even an experimental project to start with and validate the concept.